### ISSUE TITLE
Flexibility with Dataset API for version > 1.4

### ISSUE THREAD

**Issue by chaos5958** [tensorlayer/tensorlayer#253]
Tensorflow offers dataset API for building input pipeline, and it seems much faster than traditional pipeline using feed_dict. 

I open this issue to discuss which parts of Tensorlayer code should be modified to support dataset API. 
For example, Dropout layer defines placeholder inside APIs, so it has limitation to use with dataset API. 

I think this is pretty important because dataset API would be the norm way to implement tensorflow program, but I'm not sure current Tensorlayer has enough flexibility to support the new functionality (e.g. dataset API). 

I'm just beginner but one of active users of Tensorlayer, and I hope the library would get better and better.

**Comment by jorgemf** [comment]
The current API uses `tl.util.fit` in TensorFlow most functionality has been added to hooks in the session. A minimum example would be (I wrote this for the distributed API):
```
 task_spec = TaskSpec()
 # dataset is a :class:`tf.data.Dataset` with the raw data
 dataset = create_dataset()
 if task_spec is not None:
     dataset = dataset.shard(task_spec.num_workers, task_spec.shard_index)
 # shuffle or apply a map function to the new sharded dataset, for example:
 dataset = dataset.shuffle(buffer_size=10000)
 dataset = dataset.batch(batch_size)
 dataset = dataset.repeat(num_epochs)
 # create the iterator for the dataset and the input tensor
 iterator = dataset.make_one_shot_iterator()
 next_element = iterator.get_next()
 with tf.device(task_spec.device_fn()):
      # next_element is the input for the graph
      tensors = create_graph(next_element)
 with tl.DistributedSession(task_spec=task_spec,
                            checkpoint_dir='/tmp/ckpt') as session:
      while not session.should_stop():
           session.run(tensors)
```

The only issue with this code is that it will show an error after the number of epochs you set. But you can set up most things in the distributed session (http://tensorlayer.readthedocs.io/en/latest/modules/distributed.html#distributed-session-object ) and you can add more functionality with hooks: https://www.tensorflow.org/api_guides/python/train#Training_Hooks
I have already added some hooks for the distributed API and I think it can be worth to add more functionality this way http://tensorlayer.readthedocs.io/en/latest/modules/distributed.html#session-hooks

The Dataset class in TensorFlow offers a lot of flexibility, I would add our own Dataset class to TensorLayer where we trade some flexibility for something easier to use. Maybe for some uses cases as reading data from text files (for example each line contains a description of a data sample and then we use a map function to load the data) and a custom generator were the samples are generated on the fly. I have already code for this that I could add to TensorLayer. I think it is also important the dataset class is related with the distributed API as we can use things like sharding to use different data samples per worker server.

**Comment by zsdonghao** [comment]
For the `DropoutLayer`, I suggest to set `is_fix` to True to build separated graphs for training and testing (instead of using placeholder to control train/test), see this example https://github.com/zsdonghao/tensorlayer/blob/master/example/tutorial_mlp_dropout2.py

**Comment by luomai** [comment]
@jorgemf @chaos5958 I have been checking the dataset API in these days. It looks pretty intuitive and the APIs are easy to use. Given the non-invasive nature of TL, it should be pretty straightforward to use the dataset APIs interchangeably with the TL code.  I want to ask for some hints here. What are the missed API that you would like TL to provide extra wrapper function on top of the dataset?

**Comment by jorgemf** [comment]
I think shuffle, sharding and interleave can be a little confusing in distributed training, we can provide something to make it simpler for users.

**Comment by zsdonghao** [comment]
hi, people use dataset API and TL together now.
I close this issue, feel free to reopen it if you want to discuss.
